src/pure.tsx

```diff
@@ -0,0 +1,125 @@
+import React, { ReactElement, ReactNode, Suspense } from 'react'
+import { act, create, ReactTestRenderer } from 'react-test-renderer'
+import asyncUtils from './asyncUtils'
+import { cleanup, addCleanup, removeCleanup } from './cleanup'
+
+// TODO: Add better type, currently file is utilizing minimum types to work
+// TODO: Attempt to refactor code to remove ESLint disables if possible
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+type Props<T = any, R = any> = {
+  callback: (props: T) => R
+  hookProps: unknown
+  onError: CallableFunction
+  children: CallableFunction
+}
+function TestHook({ callback, hookProps, onError, children }: Props) {
+  try {
+    children(callback(hookProps))
+    // eslint-disable-next-line @typescript-eslint/no-implicit-any-catch
+  } catch (err) {
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
+    if (err.then) {
+      throw err
+    } else {
+      onError(err)
+    }
+  }
+  return null
+}
+
+function Fallback() {
+  return null
+}
+
+function resultContainer<R>() {
+  const results: Array<{ value?: R; error?: Error }> = []
+  const resolvers: Array<VoidFunction> = []
+
+  const result = {
+    get all() {
+      return results.map(({ value, error }) => error ?? value)
+    },
+    get current() {
+      const { value, error } = results[results.length - 1]
+      if (error) {
+        throw error
+      }
+      return value as R
+    },
+    get error() {
+      const { error } = results[results.length - 1]
+      return error
+    }
+  }
+
+  const updateResult = (value?: R, error?: Error) => {
+    results.push({ value, error })
+    resolvers.splice(0, resolvers.length).forEach((resolve) => resolve())
+  }
+
+  return {
+    result,
+    addResolver: (resolver: VoidFunction) => {
+      resolvers.push(resolver)
+    },
+    setValue: (value: R) => updateResult(value),
+    setError: (error: Error) => updateResult(undefined, error)
+  }
+}
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+function renderHook<T = any, R = any>(
```

Comment:
I think we can remove the default value (`any`) because it will be inferred

Reply:
Agreed that these should ideally be inferred in s many situations as possible.

I'm a bit worried about the scenario where `initialProps` is not provided in the `renderHooks` call, but new props are provided in the `rerender` call, e.g.

```ts
test('should infer from initialProps', () => {
  const { rerender } = renderHook(({ arg }) => useSomeHook(arg), {
    initialProps: { arg: false }
  })

  rerender({ arg: true })
})

test('should infer from defaulted args', () => {
  const { rerender } = renderHook(({ arg = false } = {}) => useSomeHook(arg))

  rerender({ arg: true })
})
```

(note: the inferring from the defaulted args is the ideal goal, but might not actually have enough information to infer  successfully, even with the best typings in the world)